designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,349 @@
+      reasons: [drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
+```
+
+We could also allow a semantic like this
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty 
+    expireAfter: 7d 
+    budgets:
+    - nodes: "100%" # Always allow Expiration and Emptiness to disrupt 
+      reasons: [expired, empty]
+    - nodes: 0
+```
+So in this case, we would take the minimum node count, and it would block the 100% expiration budget of expired and Empty. 
+But this semantic would not work for equation 3. Equation three will take the minimum between minNodeCount[reason] and the minNodeCount[default/unspecifed].  
+
+#### Fully Blocking Drift && Expiration while allowing consolidation
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [drifted, expired] 
+    - nodes: "100%" // All Disruption Other than drift and expiration are allowed
+```
+
+#### Fully Blocking Drift and Expiration while allowing emptiness 
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenEmpty
+    expireAfter: Never 
+    budgets:
+    - nodes: 0
+      reasons: [drifted] 
+    - nodes: "100%" // All Disruption Other than drift(Expiration and Empty Nodes) are allowed
+```
+
+#### Limiting Drift & Expiration while allowing aggressive but not unbounded consolidation
+Drift and Expiration may not be desired reasons to remove many nodes in one interval, but Underutilization and Emptiness may be desired. We may want to still consolidate at a healthy threshold but not allow 
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: worker-pool 
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    consolidationPolicy: WhenUnderutilized 
+    expireAfter: 1h 
+    budgets:
+    - nodes: 1
+      reasons: [drifted, expired] # Only allow Drift and Expiration to occur one at a time 
+    - nodes: "33%" # All Consolidation actions taken either through replace or delete should be allowed to disrupt at least a third of the cluster at once 
+```
+
+### Q: How should we handle an unspecified reason when others are specified? 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drifted, underutilized]
+    schedule: "* * * * *"
+  - nodes: 5 
+    reasons: [empty] 
+    schedule: "* * * * *" 
+```
+In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  
```

Comment:
How is the default undefined? I would assume the default is the `reasons: []` budget

Reply:
in this particular example 

```yaml
budgets: 
  - nodes: 10
    reasons: [drifted, underutilized]
    schedule: "* * * * *"
  - nodes: 5 
    reasons: [empty] 
    schedule: "* * * * *" 
```
We have a disruption budget mapping of 
budgetMapping = {
   drifted: 10 
   underutilized: 10 
   empty: 5 
}

See here we have no budget defined that covers the behavior of expired. So what should we default the values for "expired" disruption? Should we allow disruption for expired nodes? Is it unbounded since no restriction is specified?  That's what I mean. 

 The "default" language is a remnant from earlier revisions of this "default" concept I had in the original design document. In this case, I am asking what should the default value be for a reason with an undefined disruption budget.